src/types/function.rs

```diff
@@ -162,7 +173,19 @@ impl PyCFunction {
         } else {
             (std::ptr::null_mut(), std::ptr::null_mut())
         };
-        Self::internal_new_from_pointers(method_def, py, mod_ptr, module_name)
+        let (def, _destructor) = method_def
+            .as_method_def()
+            .map_err(|err| PyValueError::new_err(err.0))?;
+
+        let def = Box::into_raw(Box::new(def));
```

Comment:
This still leaks; I haven't looked at what actually calls this function so I'm not sure if that's important.

Reply:
It's called by the `PyCFunction::new` and `new_with_keywords` methods.

I think the only way around this would be to have a way to construct `&static ffi::PyMethodDef` safely and change the `PyCFunction` apis to take it (even if it's wrapped up in another type so users can't see it).

Either that, or should they be changed to use a capsule in the same way as `PyCFunction::new_closure`?

Reply:
> Either that, or should they be changed to use a capsule in the same way as PyCFunction::new_closure?

I think this would be the smallest change to fix this. Everything else could be a follow-up optimization.

(Generally speaking, I think there would be an optimization opportunity by creating a `#[pyclass]` wrapping a `ffi::PyMethodDef` instead of using a capsule to avoid one layer of dynamic allocation as `ffi::PyMethodDef` could be stored inline in the `#[pyclass]` whereas the capsule is itself limited to a pointer if I understand things correctly.)

Reply:
While true, we've steered clear of having `#[pyclass]` types internal to PyO3 because the wouldn't be publicly accessible or re-used between PyO3 modules. Solutions to that seem hard. (Maybe it's not actually a problem though.)

Reply:
> because the wouldn't be publicly accessible

I think that would not be a problem here as this should really be an implementation detail. I suspect having to manually write the `#[pyclass]` to avoid the macros as it used to be done `rust-numpy` would be a bit of a hassle though (but easier from within PyO3 than from within `rust-numpy`).

> re-used between PyO3 modules.

I guess this makes it a trade-off between less indirection but more type objects. There are usually multiple method definitions per extension module, so I suspect this could work out. At least I know of no complaints w.r.t `rust-numpy`'s `PySliceContainer` which suffers from the same issue.

(This does mean that this could be applied only for "one-way" types though, i.e. when we hand data to Python for taking care of. If we might happen to consume such a type created by a different PyO3-written extension module, this would break.)

Reply:
Actually can use the `#[pyclass(crate = "crate")]` to just use the macros I think. However it'd mean this feature wouldn't work if the macros feature was disabled.

Yes, perhaps you are right that in this "internal" case it would be suitable.